From 69ff2bdb4dd6b8388694a8e66563e1d7d93e1c25 Mon Sep 17 00:00:00 2001
From: Mohammed Rafi KC <rkavunga@redhat.com>
Date: Thu, 10 Sep 2015 14:19:06 +0530
Subject: [PATCH 08/23] Tier/glusterd: Do not allow attach-tier if remove-brick is not committed

When attaching a tier, if there is a pending remove-brick
task, then should not allow attach-tier. Since we are not
supporting add/remove brick on a tiered volume, we won't
able to commit pending remove-brick after attaching the
tier

Back port of>
>Change-Id: Ib434e2e6bc75f0908762f087ad1ca711e6b62818
>BUG: 1261819
>Signed-off-by: Mohammed Rafi KC <rkavunga@redhat.com>
>Reviewed-on: http://review.gluster.org/12148
>Tested-by: Gluster Build System <jenkins@build.gluster.com>
>Reviewed-by: Dan Lambright <dlambrig@redhat.com>
>Tested-by: Dan Lambright <dlambrig@redhat.com>

>(cherry picked from commit bc11be7864eb7f22ad6b529e95bac5a2833f5a01)

>Change-Id: I96a6085d215861663eb83a55173282a015976662
>BUG: 1258833
>Signed-off-by: Mohammed Rafi KC <rkavunga@redhat.com>
>Reviewed-on: http://review.gluster.org/12245
>Tested-by: Gluster Build System <jenkins@build.gluster.com>
>Tested-by: NetBSD Build System <jenkins@build.gluster.org>
>Reviewed-by: Dan Lambright <dlambrig@redhat.com>

Change-Id: I83ea6a67fb909eb65d27761b4a338957b618aabb
BUG: 1271725
Signed-off-by: Mohammed Rafi KC <rkavunga@redhat.com>
Reviewed-on: https://code.engineering.redhat.com/gerrit/59397
Reviewed-by: Dan Lambright <dlambrig@redhat.com>
Tested-by: Dan Lambright <dlambrig@redhat.com>
---
 xlators/mgmt/glusterd/src/glusterd-brick-ops.c |   24 ++++++++++++++++++++++++
 1 files changed, 24 insertions(+), 0 deletions(-)

diff --git a/xlators/mgmt/glusterd/src/glusterd-brick-ops.c b/xlators/mgmt/glusterd/src/glusterd-brick-ops.c
index 26ee5c5..282a78a 100644
--- a/xlators/mgmt/glusterd/src/glusterd-brick-ops.c
+++ b/xlators/mgmt/glusterd/src/glusterd-brick-ops.c
@@ -1537,6 +1537,30 @@ glusterd_op_stage_add_brick (dict_t *dict, char **op_errstr, dict_t *rsp_dict)
                 ret = -1;
                 goto out;
         }
+
+        if (dict_get(dict, "attach-tier")) {
+
+                /*
+                 * This check is needed because of add/remove brick
+                 * is not supported on a tiered volume. So once a tier
+                 * is attached we cannot commit or stop the remove-brick
+                 * task. Please change this comment once we start supporting
+                 * add/remove brick on a tiered volume.
+                 */
+                if (!gd_is_remove_brick_committed (volinfo)) {
+
+                        snprintf (msg, sizeof (msg), "An earlier remove-brick "
+                                  "task exists for volume %s. Either commit it"
+                                  " or stop it before attaching a tier.",
+                                  volinfo->volname);
+                        gf_msg (THIS->name, GF_LOG_ERROR, 0,
+                                GD_MSG_OLD_REMOVE_BRICK_EXISTS, "%s", msg);
+                        *op_errstr = gf_strdup (msg);
+                        ret = -1;
+                        goto out;
+                }
+        }
+
         ret = dict_get_int32 (dict, "count", &count);
         if (ret) {
                 gf_msg ("glusterd", GF_LOG_ERROR, errno,
-- 
1.7.1

